src/loctool/PendoXliffFile.ts

```diff
@@ -0,0 +1,446 @@
+    attributes: { datatype: typeof XLIFF.VALUE.PENDOGUIDE };
+};
+type TransUnitElement = Element & { name: typeof XLIFF.ELEMENT.TRANS_UNIT };
+
+const XLIFF = {
+    ELEMENT: {
+        XLIFF: "xliff",
+        FILE: "file",
+        BODY: "body",
+        GROUP: "group",
+        TRANS_UNIT: "trans-unit",
+        SOURCE: "source",
+        TARGET: "target",
+        NOTE: "note",
+    },
+    ATTRIBUTE: {
+        ID: "id",
+        RESNAME: "resname",
+        DATATYPE: "datatype",
+        RESTYPE: "restype",
+        SOURCE_LOCALE: "source-language",
+        TARGET_LOCALE: "target-language",
+        STATE: "state",
+        VERSION: "version",
+    },
+    VALUE: {
+        TRANSLATED: "translated",
+        PENDOGUIDE: "pendoguide",
+    },
+} as const;
+
+export class PendoXliffFile implements File {
+    // @TODO logger
+
+    /**
+     * Path to the file being localized (i.e. source file).
+     */
+    private readonly sourceFilePath: string;
+
+    /**
+     * Parsed XLIFF file.
+     *
+     * Used during extraction to obtain translation units,
+     * and later during localization to create a localized copy of it.
+     */
+    private xliffDocument?: Element;
+
+    /**
+     * Modified translation units.
+     *
+     * This is a result of extracting translation units from the XLIFF file
+     * and escaping their markdown syntax.
+     *
+     * Translation Units have modified source strings (escaped markdown syntax) and
+     * modified comments (appended component descriptions).
+     *
+     * Such escaped TUs are output to loctool for translation after extraction.
+     */
+    private escapedUnits?: TranslationUnit[];
+
+    /**
+     * Data about markdown syntax escaped in corresponding source strings.
+     *
+     * This is used to backconvert localized strings to their original markdown syntax.
+     *
+     * Null values indicate that no components were found in the source string
+     * and there is no need for backconversion.
+     */
+    private componentLists?: { [unitKey: string]: ComponentList | null };
+
+    constructor(
+        sourceFilePath: string,
+        getLocalizedPath: typeof this.getLocalizedPath,
+        getOuputLocale: typeof this.getOuputLocale,
+        createTranslationSet: typeof this.createTranslationSet,
+    ) {
+        this.sourceFilePath = sourceFilePath;
+        this.getLocalizedPath = getLocalizedPath;
+        this.getOuputLocale = getOuputLocale;
+        this.createTranslationSet = createTranslationSet;
+    }
+
+    private static loadXliffDocument(path: string) {
+        const content = fs.readFileSync(path, "utf-8");
+
+        // parse XML
+        const document = xml2js(content, { compact: false }) as Element;
+
+        // ensure it's a version 1.2 XLIFF
+        const root = PendoXliffFile.getXliffRoot(document);
+        if (root.attributes?.[XLIFF.ATTRIBUTE.VERSION] !== "1.2") {
+            throw new Error("Only XLIFF version 1.2 is supported");
```

Comment:
This seems to be critical - I do not see test for this

Reply:
Added 446aa20